ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -371,44 +390,96 @@ public void replicate() {
+              getRemoteStandbyReplicasTimedOutOnNoProgress(replicasToReplicatePerNode);
+          if (standbyReplicasTimedOutOnNoProgress.size() > 0) {
+            allCaughtUp = false;
+            currentReplicaList = standbyReplicasTimedOutOnNoProgress;
+            if (connectedChannel == null) {
+              connectedChannel = connectionPool.checkOutConnection(remoteNode.getHostname(),
+                  standbyReplicasTimedOutOnNoProgress.get(0).getPort(),
+                  replicationConfig.replicationConnectionPoolCheckoutTimeoutMs);
+              checkoutConnectionTimeInMs = SystemTime.getInstance().milliseconds() - startTimeInMs;
+            }
+            List<ExchangeMetadataResponse> exchangeMetadataResponseListForInactiveReplicas =
+                standbyReplicasTimedOutOnNoProgress.stream()
+                    .map(RemoteReplicaInfo::getExchangeMetadataResponse)
+                    .collect(Collectors.toList());
+            exchangeMetadataTimeInMs = 0;
+            fixMissingStoreKeysTimeInMs = -1;
+            logger.debug(
+                "Sending GET request to fetch missing keys for standby remote replicas {} timed out on no progress",
+                currentReplicaList);
+            fixMissingStoreKeys(connectedChannel, standbyReplicasTimedOutOnNoProgress,
+                exchangeMetadataResponseListForInactiveReplicas);
             fixMissingStoreKeysTimeInMs = SystemTime.getInstance().milliseconds() - startTimeInMs;
           }
-          logger.error(
-              "Error while talking to peer: Remote node: {}, Thread name: {}, Remote replicas: {}, Current active "
-                  + "remote replica list: {}, Checkout connection time: {}, Exchange metadata time: {}, Fix missing "
-                  + "store key time {}", remoteNode, threadName, replicasToReplicatePerNode, currentReplicaList,
-              checkoutConnectionTimeInMs, exchangeMetadataTimeInMs, fixMissingStoreKeysTimeInMs, e);
-          replicationMetrics.incrementReplicationErrors(replicatingOverSsl);
-          if (connectedChannel != null) {
-            connectionPool.destroyConnection(connectedChannel);
-            connectedChannel = null;
-          }
-        } finally {
-          long totalReplicationTime = SystemTime.getInstance().milliseconds() - replicationStartTimeInMs;
-          replicationMetrics.updateTotalReplicationTime(totalReplicationTime, replicatingFromRemoteColo,
-              replicatingOverSsl, datacenterName);
-          if (connectedChannel != null) {
-            connectionPool.checkInConnection(connectedChannel);
-          }
-          context.stop();
-          portTypeBasedContext.stop();
         }
+      } catch (Throwable e) {
+        if (checkoutConnectionTimeInMs == -1) {
+          // throwable happened in checkout connection phase
+          checkoutConnectionTimeInMs = SystemTime.getInstance().milliseconds() - startTimeInMs;
```

Comment:
minor: we probably can replace all `SystemTime.getInstance()` with `time` in this file.